### PR TITLE
 Buildings code rewrite

### LINKED ISSUE #2532: Rethinking major and minor buildings
Rewriting the buildings in #565 took place 1,5 year ago and proved to be right in general, but it's time to fine tune it. It become apparent to me in a discussion about unifying current major buildings (https://github.com/gravitystorm/openstreetmap-carto/pull/2515), which I meant as a simple, clearly focused fix, but turned into discussing special buildings rendering.

There are some related issues already created:
- https://github.com/gravitystorm/openstreetmap-carto/issues/738 (Render building=greenhouse differently)
- https://github.com/gravitystorm/openstreetmap-carto/issues/1206 (Amenity=place_of_worship buildings)
- https://github.com/gravitystorm/openstreetmap-carto/issues/1207 (building=house and building=garage should render in slightly different colours)

but they are also related to certain features, while the underlying problem is wider:
1. Do we want having any other building types rendering than just standard brown?
2. If yes, which buildings should be rendered as major?
3. If yes, which buildings should be rendered as minor?

My take is that it's good to have basic subtypes of building class which is very large and important for universal map style, but has no clear rules for differentiation. That means it's up to us to decide. Some buildings are landmarks or plays a special function (like places of worship or transportation terminals), while the others are less important than typical buildings (like greenhouses, sheds, roofs or garages) and we should show it if possible. Major building list should be however treated with more care than minor ones to [avoid visual clutter](https://github.com/gravitystorm/openstreetmap-carto/pull/2515#issuecomment-270024246).

My opinion is that using dark brown for major buildings and light brown for minor ones is a consistent rule and hints the users that all these are still buildings.

### PR DESCRIPTION
**Pull request by meased**
(Most likely) a prerequisite for #2532

This PR simplifies the buildings SQL query and moves all styling decisions out of SQL and into CSS (as the software gods intended :wink:).

No visual changes.

## SQL

Before, normal buildings and major building were queried separately with each query specifically omitting the results of the other. This made it difficult to redefine the categories (and very difficult to add a third category).

The SQL query now queries all buildings together, along with any fields needed to make styling decisions:

 ``` sql
      (SELECT
            way,
            building,
            amenity,
            aeroway,
            aerialway,
            tags->'public_transport' as public_transport
          FROM planet_osm_polygon
          WHERE building IS NOT NULL
            AND building != 'no'
            AND way_area > 1*!pixel_width!::real*!pixel_height!::real
          ORDER BY COALESCE(layer,0), way_area DESC
        ) AS buildings
```

## CSS

Now that the buildings class has access to all the fields needed, the CSS can be written in one block. As you can see, it is now very easy to reclassify buildings and add/remove the minor class:

``` css
  #buildings {
    [zoom >= 13] {
      polygon-fill: @building-low-zoom;
      polygon-clip: false;
      line-width: 0;
      [zoom >= 15] {
        polygon-fill: @building-fill;
        line-color: @building-line;
        line-width: .75;
        line-clip: false;
      }
      [amenity = 'place_of_worship'],
      [aeroway = 'terminal'],
      [aerialway = 'station'],
      [building = 'train_station'],
      [public_transport = 'station'] {
        polygon-fill: @building-major-fill;
        line-color: @building-major-line;
      }
    }
  }
```

### CODE REVIEW

**Comment by jeisenbe** [comment]
Great work on the code!

I think it might be better with the major buildings a little more
prominent.

Could we try a color half-way between the old and new major building colors?

**Comment by kocio-pl** [comment]
I like the idea of these complex changes. Did you however test the performance? We use SQL a lot to achieve reasonable speed.

**Comment by kocio-pl** [comment]
Could you also test minor buildings darker, like half the way between the proposed color and standard buildings?

**Comment by meased** [comment]
Minor buildings no change:
0perc

Minor buildings 4% lighter:
4perc

Minor buildings 7% lighter (same as screen shots above):
7perc

**Comment by meased** [comment]
Major buildings no change:
maj2_20perc
maj_20perc

Major buildings 15% darker:
maj2_15perc
maj_15perc

Major buildings 10% darker (same as screenshots above):
maj2_10perc
maj_10perc

As you can see, I was more concerned with fixing the dreadful "gastronomy in the airport" problem than I was with spotting churches in neighbourhoods.

**Comment by meased** [comment]
> Did you however test the performance?

No.

Anyone know how to run benchmarks in docker?

**Comment by jeisenbe** [comment]
Would it make sense to lighten major buildings at higher zoom levels? At
mid-zoom it is very helpful that the airport terminal building is obvious,
and that’s when it is useful for places of worship etc.

But at zoom 18 there is no reason for the major building to stay so dark.

Would that be too complicated ?

**Comment by meased** [comment]
That's not that complicated. Worth a test rendering at least.

**Comment by kocio-pl** [comment]
I think both darker versions (4% for minor and 15% for major) work better. Since buildings are very popular map feature, we need much more testing in different areas (backgrounds). 

I like the idea of minor buildings instead of problems with ruins icon, especially for castles (see #331).

Unfortunately we have no performance framework (see #1287). However we can at least try to deploy it on some real server - @rrzefox, could you measure the effect of replacing some SQL code with Carto CSS here?

**Comment by matthijsmelissen** [comment]
See also #68 and #490.

I'm not sure if I really see the advantage of rendering minor buildings lighter.

**Comment by mboeringa** [comment]
> I'm not sure if I really see the advantage of rendering minor buildings lighter.

That surprises me. To me it makes utter sense to render relatively insignificant non-living quarters like small garden sheds and garages lighter.

It might actually stimulate good tagging practices in that people actually start to distinguish them in tagging, and add **building=shed** instead of just generic **building=yes**.

**Comment by Tomasz-W** [comment]
@meased Can you test some lighter outline for minor buildings? It looks a little bit like some holes in ground currently.

**Comment by mboeringa** [comment]
> @meased Can you test lighter outline for minor buildings? It looks a little bit like some holes in ground currently.

I think that is mainly a consequence of the relatively dark grey of the **landuse=residential**, which appears just a bit darker than the minor buildings. So either slightly lightening up of landuse=residential, or slightly darkening the minor buildings fill, might help.

**Comment by meased** [comment]
> I think that is mainly a consequence of the relatively dark grey of the landuse=residential

Exactly.

I choose landuse=residential and airports for examples because they represent the worst cases for how light you can go on one end and how dark you can go on the other. There is not much space to play with in the middle. The minor buildings above would have looked better on bare background:
base_minor

**Comment by kocio-pl** [comment]
> It might actually stimulate good tagging practices in that people actually start to distinguish them in tagging, and add building=shed instead of just generic building=yes.

I also believe it's more pleasant, less cluttering and more useful at the same time, so it's a triple win for me.

There are a lot of auxiliary buildings, but still most of them is tagged with `building=yes` (and only about 50 mln out of 300 mln of all the buildings has any other value!). I don't see how could we attract people to more precise tagging other than showing them a bit different as a "minor" super-class.

taghistory 55
taghistory 56

@meased I see that you have added carpark type, but omitted collapsed - was there any reason for this? I have tried to stay above 10k limit, but we don't have to stick to it, there are also some other types that make sense for me, what do you think about it?

taghistory 57

**Comment by jragusa** [comment]
Please also consider underground #552 and roof #2475 buildings to get a consistent rendering.

**Comment by meased** [comment]
I got the original minor buildings list from #2532. I was avoiding underground (because some people don't want it rendered at all) and roof (because #2475 is different than just rendering them lighter). But until those other issues are sorted out I don't see why they wouldn't qualify as minor buildings in the meantime.

@kocio-pl: Strangly, collapsed isn't on the [wiki page](https://wiki.openstreetmap.org/wiki/Key:building). I included it anyway because the usage is very high.

The list of minor buildings is now: 

``` css
    [building = 'roof'],
    [building = 'construction'],
    [building = 'damaged'],
    [building = 'collapsed'],
    [building = 'underground'],
    [building = 'service'],
    [building = 'slurry_tank'],
    [building = 'shed'],
    [building = 'garage'],
    [building = 'garages'],
    [building = 'farm_auxiliary'],
    [building = 'carport'],
    [building = 'barn'],
    [building = 'stable'],
    [building = 'cowshed'],
    [building = 'greenhouse'],
    [building = 'ger'],
    [building = 'ruins'] {
```

**Comment by meased** [comment]
Playing with @jeisenbe's idea of variable major building darkness:

a14
a15
a16
a17


c14
c15
c16
c17

**Comment by Tomasz-W** [comment]
About test renderings above: I find darkness of major buildings on high zoom levels definetely too low. When I looked at them with "fresh eye" I had to search for major buildings, but they should be distinguishable intuitively. Just to be clear: I very like the idea proposed in this PR, it just need some tuning IMO.

File: buildings.mss [hunk 1]
```diff
@@ -12,29 +15,41 @@
...
+    [building = 'shed'],
+    [building = 'garage'],
+    [building = 'garages'],
+    [building = 'farm_auxiliary'],
+    [building = 'carport'],
+    [building = 'barn'],
+    [building = 'stable'],
+    [building = 'cowshed'],
+    [building = 'greenhouse'],
+    [building = 'ger'],
```
Comment (HolgerJeromin) [inline buildings.mss]: What is that?

I miss =hut and perhaps =cabin in the list.

Comment (kocio-pl) [reply buildings.mss]: You mean ger? See https://en.wikipedia.org/wiki/Yurt

**Comment by kocio-pl** [comment]
What about construction or ruins?

**Comment by dieterdreist** [comment]
sent from a phone

> On 9. Oct 2018, at 03:58, kocio-pl <notifications@github.com> wrote:
> 
> What about construction or ruins?


I’m undecided, they are not building types but they work like elsewhere, e.g. highway=construction, construction=primary

And they are at least buildings

**Comment by kocio-pl** [comment]
@meased Variable darkness is smart trick and it can work for me. More tests would be useful to make sure, because buildings are very popular. 

The most important think for now is deciding what the list should contain. I like the idea of tagging `building=construction/ruins/collapsed/damaged` + `construction/ruins/collapsed/damaged=*`, but I'm not sure what should we render then - their uncomplete state as lighter or just their type.

***

Example of gers (yurts/tents) being less visible than regular buildings (in this area some of them are not tagged properly, so the difference in rendering is evident), using current code:

https://www.openstreetmap.org/#map=18/47.93434/106.91192

ft9dvqo1

**Comment by rrzefox** [comment]
> However we can at least try to deploy it on some real server - @rrzefox, could you measure the effect of replacing some SQL code with Carto CSS here?

I will gladly deploy this once it has reached "beta" state, i.e. coloring has been decided.
I seriously doubt this has any measureable performance impact.

**Comment by Adamant36** [comment]
@rrzefox, since we are on the subject of performance, do you know what impact there would be from creating more stylesheets with what is currently in the amenity file?

**Comment by Tomasz-W** [comment]
@meased Can you try with 17% for major and 5% for minor due to some users questions about further tests? It would be good to see some new locations to have a 'fresh view' on it.

**Comment by meased** [comment]
No problem, I'll do more testing this weekend.

**Comment by kocio-pl** [comment]
I think deploying this idea will help people realize that there are many different building types and start adding them more often. In the looooong term this might lead to render `building=yes` as kind of minor/incomplete. So this is much more to this change than meets the eye!

I believe in such incremental, but more systematic changes. For example making more categories for POIs, as I did purely for making this style easier to customize, has led to making gastronomy icons orange, which may help to avoid clutter by starting rendering them as dots.

**Comment by dieterdreist** [comment]
sent from a phone

> On 12. Oct 2018, at 19:29, kocio-pl <notifications@github.com> wrote:
> 
> I think deploying this idea will help people realize that there are many different building types and start adding them more often. In the looooong term this might lead to render building=yes as kind of minor/incomplete. So this is much more to this change than meets the eye!


+1, well said, the times that mapping projects called themselves proudly building=yes should be gone ;-)

**Comment by meased** [comment]
I have removed `collapsed` and added `hut` and `cabin`.

I have left `ruins` off due to similarity to `collapsed`.

Still undecided on `construction`. I feel it almost warrants a new style altogether, which would be better implemented in a separate PR, so I've left it off for now.

There is a lot going on in this PR so I am thinking about making this PR only about adding minor buildings (and the query rewrite) and making a separate PR later for lightening major buildings. I think this would help get things done faster as we will be able to focus on one thing at a time. Unless anyone objects, I'll update the repository and change the title of this PR.

**Comment by meased** [comment]
More testing (at 7% lighter):

https://www.openstreetmap.org/#map=17/45.58231/-122.73927
sj_b
sj_a

https://www.openstreetmap.org/#map=17/45.15308/-122.84857
wb_b
wb_a

https://www.openstreetmap.org/#map=18/45.46846/-122.73448
hv_b
hv_a

https://www.openstreetmap.org/#map=17/45.50819/-122.64908
la_b
la_a

aly_b
aly_a

**Comment by meased** [comment]
@Tomasz-W wanted to see 5% lighter:

0%
la1_0p

5%
la1_5p

7%
la1_7p

0%
la2_0p

5%
la2_5p

7%
la2_7p

**Comment by Tomasz-W** [comment]
Thanks! As 5% is too dark to distinguish them fastly, but 7% is minimally too light (too prominent) on residential areas, I would go with middle one of 6%.

**Comment by meased** [comment]
6% version of the last image:
la2_6p

**Comment by jeisenbe** [comment]
Looks good on residential landuse.

Can we try an example with minor buildings on Religious landuse?

**Comment by Adamant36** [comment]
@meased, can you try a couple of tests with 6% and the new yellowish school grounds color?

**Comment by kocio-pl** [comment]
I think it's important to test it on different backgrounds where one can find buildings (for example `landuse=garages` or `landuse=commercial`), not just the bare ground. They will look different there, so it's too early for precise shade tuning.

**Comment by kocio-pl** [comment]
@meased My proposition is to start with just the code restructuring as a separate PR, so testing would be easier and less error prone, and next making another PR based on this which will be focused only on making minor category and visual testing the colors.

**Comment by meased** [comment]
> My proposition is to start with just the code restructuring as a separate PR

I think I'll do this.

**Comment by meased** [comment]
Done.

This PR is now just a rewrite of the buildings SQL and style sheet. There should be no visual differences. I have tested this in Kosmtik by flipping the overlay of the standard map on and off and I can't see any difference on any zoom level.

**Comment by kocio-pl** [comment]
Great! Since there is a change from SQL code to CSS code, there might be some performance difference. @rrzefox Could you test this difference?

**Comment by kocio-pl** [comment]
Well, this statement was not accurate - it's rather simpler both SQL and CSS code, so there should be no problems with performance. I will just wait with this until v4.16.0 is out (I plan to do this 19.10) and probably merge this PR soon after that. Then we can start testing minor buildings PR.

**Comment by kocio-pl** [comment]
BTW: making major buildings lighter would also fix #3071.

**Comment by kocio-pl** [comment]
@meased What about further tuning? Do you plan to get back to the proposed minor and major buildings changes?

**Comment by meased** [comment]
The problem here is that buildings need to simultaneously serve as a background for our lightest foreground (I believe that is gastromony `#C77400`) and as a foreground to our darkest landusage (which I believe is religous `#d0d0d0`, but residential is also a problem), all while having three brightnesses of minor, normal, and major which are all distinguishable from each other.

I'm not convinced that all of these criteria can be met, hence the inactivity.

Regardless, I could start a PR so that these problems could be investigated.

**Comment by matthijsmelissen** [comment]
That's the kind of thinking we need! Feel free to propose dropping/altering any of these criteria if you think it would make sense.

By the way, just curious, what's your username on openstreetmap.org?

**Comment by meased** [comment]
> what's your username on openstreetmap.org?

[cowdog](https://www.openstreetmap.org/user/cowdog)

**Comment by Tomasz-W** [comment]
Please remember about discussed half-transparency for some buildings (https://github.com/gravitystorm/openstreetmap-carto/issues/552#issuecomment-446540473).

**Pull request merged**